test/e2e/canary_test.go

```diff
@@ -160,8 +212,10 @@ spec:
 		// See: https://github.com/argoproj/argo-rollouts/issues/738
 		ExpectCanaryStablePodCount(6, 4).
 		When().
-		ScaleRollout(4)
-	// WaitForRolloutReplicas(4) // this doesn't work yet (bug)
+		ScaleRollout(4).
+	    WaitForRolloutReplicas(6).
+		Then().
+		ExpectCanaryStablePodCount(4, 2)
```

Comment:
I still think this is not correct. According to the sequence of events:

1. At T-0, we were at: 6 pods (2 canary, 4 stable, 4 available)
2. At T-1, we had a scale-up event and went to: 10 pods (6 canary, 4 stable, 4 available).
3. At T-2, we had a scale-down event and went back down to 6 pods (4 canary, 2 stable, 2 available)

Notice that at T-2, we have now violated maxUnavailable because we only have 2 pods available when the spec (maxUnavailable: 0) requires 4 to be available. 

Although this is a day-0 issue, I think the proper fix may be to address https://github.com/argoproj/argo-rollouts/issues/738. If we fix https://github.com/argoproj/argo-rollouts/issues/738, then this problem may also be fixed.

Reply:
Yes, you are correct that it requires changes to adhere to both maxSurge and minAvailabile. But bug referred in #738 is to fix weights inline to deployment.  The current I am trying to fix is not scaleDown completely on either stable or canary which I will address as a part of this PR and will take #738 to do refactor to fix the whole calculation.

Reply:
Can you label #738 to 1.2?

Reply:
> But bug referred in #738 is to fix weights inline to deployment

The two are closely related, if not the same issue. The behavior you saw in #1415 is a manifestation of the same underlying problem but with different parameters of maxSurge/minAvailable. 

This fix appears to addressing the part of symptom, whereas #738 is actually the root problem. I believe if we fix #738, then the incident in #1415 would not have happened.

I do recognize that this fix improves the current behavior, but I'm pretty sure the code in this PR will need to be thrown away in order to address #738

Reply:
Agree, #738 would. be the fix we need to act on. My concern currently is current fix will stop the downtime in case of bad image versions. 
I wanted to take #738 for 1.2 to refactor the calculation and I don't want to refactor at the current stage given the complexity.